src/dune_rpc_impl/server.ml

```diff
@@ -298,14 +305,45 @@ let handler (t : t Fdecl.t) : 'a Dune_rpc_server.Handler.t =
       let* running = Fiber.Pool.running t.pool in
       match running with
       | false -> Fiber.return ()
-      | true ->
-        let errors = Build_system.errors () in
-        Fiber.Pool.task t.pool ~f:(fun () ->
-            let events =
-              List.map errors ~f:(fun (e : Build_system.Error.t) ->
-                  Diagnostic.Event.Add (diagnostic_of_error e))
-            in
-            Session.notification session Server_notifications.diagnostic events)
+      | true -> (
+        match sub with
+        | Subscribe.Diagnostics ->
+          let errors = Build_system.errors () in
+          Fiber.Pool.task t.pool ~f:(fun () ->
+              let events =
+                List.map errors ~f:(fun (e : Build_system.Error.t) ->
+                    Diagnostic.Event.Add (diagnostic_of_error e))
+              in
+              Session.notification session Server_notifications.diagnostic
+                events)
+        | Build_progress ->
+          let current_progress : Build_system.Progress.t =
+            Build_system.get_current_progress ()
+          in
+          let progress : Dune_rpc_private.Progress.t =
+            let complete = current_progress.number_of_rules_executed in
+            { complete
+            ; remaining = current_progress.number_of_rules_discovered - complete
+            }
+          in
+          let last_event =
+            match Build_system.last_event () with
+            | Some evt -> rpc_event_of_build_event evt
+            | None ->
+              (* cwong: This is technically a lie, but the alternative is
+                 minting a new [Not_started] variant that is not substantively
+                 different *)
```

Comment:
I think I'd prefer the `Not_started` or `Waiting` as opposed to this white lie. It is not urgent though.

Reply:
I've added the new event.